commit 05faae5d0ed6b004724cb902b0f9f86fc24146aa
Author: Susan Franklin <sfranklin>
Date:   Wed Oct 15 22:33:22 2008 +0000

    Bug 251015 - [ui] "could not determine what to do" while checking for updates

2	0	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/PlanStatusHelper.java
4	1	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/InstallAction.java
9	1	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/ProfileModificationAction.java
16	1	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/UninstallAction.java
18	3	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/UpdateAction.java
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/PlanStatusHelper.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/PlanStatusHelper.java
index e1c4436..b33ca88 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/PlanStatusHelper.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/PlanStatusHelper.java
@@ -45,6 +45,8 @@ public class PlanStatusHelper {
 				return new Status(IStatus.ERROR, ProvUIActivator.PLUGIN_ID, statusCode, NLS.bind(ProvUIMessages.PlanStatusHelper_NothingToDo, getIUString(affectedIU)), null);
 			case IStatusCodes.PROFILE_CHANGE_ALTERED :
 				return new MultiStatus(ProvUIActivator.PLUGIN_ID, statusCode, ProvUIMessages.PlanStatusHelper_RequestAltered, null);
+			case IStatusCodes.NOTHING_TO_UPDATE :
+				return new Status(IStatus.INFO, ProvUIActivator.PLUGIN_ID, IStatusCodes.NOTHING_TO_UPDATE, ProvUIMessages.UpdateOperation_NothingToUpdate, null);
 			default :
 				return new Status(IStatus.ERROR, ProvUIActivator.PLUGIN_ID, statusCode, NLS.bind(ProvUIMessages.PlanStatusHelper_UnexpectedError, new Integer(statusCode), getIUString(affectedIU)), null);
 		}
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/InstallAction.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/InstallAction.java
index a2faf44..4f4c108 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/InstallAction.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/InstallAction.java
@@ -99,10 +99,13 @@ public class InstallAction extends ProfileModificationAction {
 			return false;
 		// We allow non-IU's to be selected at this point, but there
 		// must be at least one installable unit selected that is
-		// not a category and is not nested underneath another IU.
+		// selectable
 		for (int i = 0; i < selectionArray.length; i++) {
 			if (selectionArray[i] instanceof InstalledIUElement && isSelectable((IUElement) selectionArray[i]))
 				return true;
+			IInstallableUnit iu = (IInstallableUnit) ProvUI.getAdapter(selectionArray[i], IInstallableUnit.class);
+			if (iu != null && isSelectable(iu))
+				return true;
 		}
 		return false;
 	}
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/ProfileModificationAction.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/ProfileModificationAction.java
index 5a7c3b3..9c4af9f 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/ProfileModificationAction.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/ProfileModificationAction.java
@@ -53,7 +53,7 @@ public abstract class ProfileModificationAction extends ProvisioningAction {
 		String id = getProfileId(true);
 		// We could not figure out a profile to operate on, so return
 		if (id == null || ius.length == 0) {
-			ProvUI.reportStatus(new Status(IStatus.ERROR, ProvUIActivator.PLUGIN_ID, ProvUIMessages.ProfileModificationAction_NoProfileToModify), StatusManager.SHOW);
+			ProvUI.reportStatus(PlanStatusHelper.getStatus(IStatusCodes.NOTHING_TO_UPDATE, null), StatusManager.BLOCK);
 			runCanceled();
 			return;
 		}
@@ -179,6 +179,10 @@ public abstract class ProfileModificationAction extends ProvisioningAction {
 				IUElement element = (IUElement) elements.get(i);
 				if (isSelectable(element))
 					iusList.add(getIU(element));
+			} else {
+				IInstallableUnit iu = (IInstallableUnit) ProvUI.getAdapter(elements.get(i), IInstallableUnit.class);
+				if (iu != null && isSelectable(iu))
+					iusList.add(iu);
 			}
 		}
 		return (IInstallableUnit[]) iusList.toArray(new IInstallableUnit[iusList.size()]);
@@ -188,6 +192,10 @@ public abstract class ProfileModificationAction extends ProvisioningAction {
 		return !(element instanceof CategoryElement);
 	}
 
+	protected boolean isSelectable(IInstallableUnit iu) {
+		return !ProvisioningUtil.isCategory(iu);
+	}
+
 	protected LicenseManager getLicenseManager() {
 		return policy.getLicenseManager();
 	}
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/UninstallAction.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/UninstallAction.java
index 651213a..47dc6d5 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/UninstallAction.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/UninstallAction.java
@@ -13,6 +13,7 @@ package org.eclipse.equinox.internal.provisional.p2.ui.actions;
 
 import org.eclipse.core.runtime.*;
 import org.eclipse.equinox.internal.p2.ui.ProvUIMessages;
+import org.eclipse.equinox.internal.p2.ui.model.IUElement;
 import org.eclipse.equinox.internal.provisional.p2.director.ProfileChangeRequest;
 import org.eclipse.equinox.internal.provisional.p2.director.ProvisioningPlan;
 import org.eclipse.equinox.internal.provisional.p2.engine.IProfile;
@@ -61,7 +62,9 @@ public class UninstallAction extends ProfileModificationAction {
 						return false;
 					}
 				} else {
-					return false;
+					IInstallableUnit iu = (IInstallableUnit) ProvUI.getAdapter(selectionArray[i], IInstallableUnit.class);
+					if (iu == null || !isSelectable(iu))
+						return false;
 				}
 			}
 			return true;
@@ -69,6 +72,18 @@ public class UninstallAction extends ProfileModificationAction {
 		return false;
 	}
 
+	protected boolean isSelectable(IUElement element) {
+		return super.isSelectable(element) && !(element.getParent(element) instanceof IUElement);
+	}
+
+	protected boolean isSelectable(IInstallableUnit iu) {
+		if (!super.isSelectable(iu))
+			return false;
+		IProfile profile = getProfile(false);
+		int lock = getLock(profile, iu);
+		return ((lock & IInstallableUnit.LOCK_UNINSTALL) == IInstallableUnit.LOCK_NONE);
+	}
+
 	protected String getTaskName() {
 		return ProvUIMessages.UninstallIUProgress;
 	}
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/UpdateAction.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/UpdateAction.java
index 6459ea8..ded2fda 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/UpdateAction.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/actions/UpdateAction.java
@@ -13,9 +13,10 @@ package org.eclipse.equinox.internal.provisional.p2.ui.actions;
 
 import java.util.*;
 import org.eclipse.core.runtime.*;
-import org.eclipse.equinox.internal.p2.ui.ProvUIActivator;
+import org.eclipse.equinox.internal.p2.ui.PlanStatusHelper;
 import org.eclipse.equinox.internal.p2.ui.ProvUIMessages;
 import org.eclipse.equinox.internal.p2.ui.model.AvailableUpdateElement;
+import org.eclipse.equinox.internal.p2.ui.model.IUElement;
 import org.eclipse.equinox.internal.provisional.p2.director.ProfileChangeRequest;
 import org.eclipse.equinox.internal.provisional.p2.director.ProvisioningPlan;
 import org.eclipse.equinox.internal.provisional.p2.engine.IProfile;
@@ -85,7 +86,7 @@ public class UpdateAction extends ProfileModificationAction {
 			sub.worked(1);
 		}
 		if (toBeUpdated.size() <= 0) {
-			status.add(new Status(IStatus.INFO, ProvUIActivator.PLUGIN_ID, IStatusCodes.NOTHING_TO_UPDATE, ProvUIMessages.UpdateOperation_NothingToUpdate, null));
+			status.add(PlanStatusHelper.getStatus(IStatusCodes.NOTHING_TO_UPDATE, null));
 			sub.done();
 			return null;
 		}
@@ -130,7 +131,9 @@ public class UpdateAction extends ProfileModificationAction {
 						return false;
 					}
 				} else {
-					return false;
+					IInstallableUnit iu = (IInstallableUnit) ProvUI.getAdapter(selectionArray[i], IInstallableUnit.class);
+					if (iu == null || !isSelectable(iu))
+						return false;
 				}
 			}
 			return true;
@@ -138,6 +141,18 @@ public class UpdateAction extends ProfileModificationAction {
 		return false;
 	}
 
+	protected boolean isSelectable(IUElement element) {
+		return super.isSelectable(element) && !(element.getParent(element) instanceof IUElement);
+	}
+
+	protected boolean isSelectable(IInstallableUnit iu) {
+		if (!super.isSelectable(iu))
+			return false;
+		IProfile profile = getProfile(false);
+		int lock = getLock(profile, iu);
+		return ((lock & IInstallableUnit.LOCK_UPDATE) == IInstallableUnit.LOCK_NONE);
+	}
+
 	protected String getTaskName() {
 		return ProvUIMessages.UpdateIUProgress;
 	}
